Section on WIMSE Identity (#7)
yaronf/wimse-s2s: draft-sheffer-wimse-s2s-protocol.md

{::boilerplate bcp14-tagged}

# Workload Identity {#whimsical-identity}

> This document defines a workload identity as a URI {{!RFC3986}}. This URI is used in the subject fields in the certificates and tokens defined later in this document. This specification treats the URI as opaque. The format of the URI and the namespace for the URI are chosen by the issuer of the token. Other specifications may define specific URI structures for particular use cases. An example of a defined identity format is the SPIFFE ID [SPIFFE-ID](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md).

Reviewer: I'm rather unsure of the wording or terminology but the URI format/namespace/etc is chosen by a higher level entity than the issuer of the token. In the WPT http://www.sheffer.org/wimse-s2s/draft-sheffer-wimse-s2s-protocol.html#section-3.2-2.2.2.1.1 case, for example, the issuer of that token is the workload itself, which has no say over the URIs used in the system.  Even the "Identity Server" is probably given a URI as it's identity in the over all system.
```suggestion
This document defines a workload identity as a URI {{!RFC3986}}. This URI is used in the subject fields in the certificates and tokens defined later in this document. This specification treats the URI as opaque. The format of the URI and the namespace for the URI are at the discretion of the deployment at large. Other specifications may define specific URI structures for particular use cases. An example of a defined identity format is the SPIFFE ID [SPIFFE-ID](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md).
```

Reviewer: Unresolving conversation as I believe the suggested change or something similar is still needed.

> A workload identity only has meaning within the scope of a specific issuer. Two identities of the same value issued by different issuers may or may not refer to the same workload. In order to avoid collisions identity URIs should include a concept of a trust domain associated with an issuer that is selected from a global name space such as host domains. However, the validator of an identity credential MUST make sure that the are using the correct issuer credential to verify the identity credential and that the issuer is trusted to issue tokens for the defined trust domain.

Author: Accepted your comment.  I'll create an issue for your other topic.

> A workload identity only has meaning within the scope of a specific issuer. Two identities of the same value issued by different issuers may or may not refer to the same workload. In order to avoid collisions identity URIs should include a concept of a trust domain associated with an issuer that is selected from a global name space such as host domains. However, the validator of an identity credential MUST make sure that the are using the correct issuer credential to verify the identity credential and that the issuer is trusted to issue tokens for the defined trust domain.

Author: Applied suggestion.

{::boilerplate bcp14-tagged}

> # WIMSE Identity

Reviewer: facilitate future referenceability a la https://github.com/yaronf/wimse-s2s/pull/8/files#r1628701212
```suggestion
# WIMSE Identity {#whimsical-identity}
```

[TODO: using a URI as an identity aligns or perhaps conflicts with SPIFFE's definition of Identity the format below is basically taken from SPIFFE, I'm not convinced that this is a good or bad idea.  Perhaps we can reuse all or most of the [SPIFFE format](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md)?]

THe WIMSE identity is represented as a URI of the format:

scheme://trustdomain/path

> The scheme is TBD, likely candidates are "spiffe" to align with SPIFFE or "wimse" provide an alternative. A particular deployment may use one or the other scheme, but not both at the same time since SPIFFE only allows one URI SAN in a certificate.

Author: Yea we can probably relax that restriction. It should probably be only one can be in any particular cert.

> The scheme is TBD, likely candidates are "spiffe" to align with SPIFFE or "wimse" provide an alternative. A particular deployment may use one or the other scheme, but not both at the same time since SPIFFE only allows one URI SAN in a certificate.

Reviewer 2: +1 that workloads can be part of multiple trust domains but SVIDs (in this case X509 certs) are only part of one.

> The scheme is TBD, likely candidates are "spiffe" to align with SPIFFE or "wimse" provide an alternative. A particular deployment may use one or the other scheme, but not both at the same time since SPIFFE only allows one URI SAN in a certificate.

Author: updated

THe trust domain is a locally defined string that is compliant with SPIFFE naming schemes. The issuer of WIMSE credentials is tied to a single trust domain. The validator of a WIMSE ID token or certificate MUST check that the issuer of the credential is permitted to issuer credentials with the specified trust domains and MUST validate that there are no collisions in the list of supported trust domains.

> The path is a string whose format is defined by the local deployment and is subject to the requirements defined in SPIFFE.  The purpose of the path is to identify a workload for the purposes of authorization, auditing and binding to additional information.

Author: We definitely should reference the spec.  I tried to limit what was discussed here, but if people do not think it adds anything it can be removed. .

> The path is a string whose format is defined by the local deployment and is subject to the requirements defined in SPIFFE.  The purpose of the path is to identify a workload for the purposes of authorization, auditing and binding to additional information.

Author: Added a link to the spec, not sure what the stable ling to SPIFFE spec is, will add a reference when we get clarity.

> The path is a string whose format is defined by the local deployment and is subject to the requirements defined in SPIFFE.  The purpose of the path is to identify a workload for the purposes of authorization, auditing and binding to additional information.

Reviewer 2: Should we add here the fact that the domain part is not governed by a central authority like it is done in DNS?

Author: I modified the text here.  I think this is probably an area for expanded discussion.
